draco/util/tools.py

```diff
@@ -534,3 +539,152 @@ def window_generalised(x, window="nuttall"):
+
+    N = y.shape[0]
+
+    if mask is None:
+        mask = np.zeros(N, dtype=bool)
+    elif np.all(mask):
+        warnings.warn("Entire dataset is masked.")
+
+        return np.zeros_like(y)
+
+    mask = np.squeeze(mask)
+
+    if mask.ndim != 1:
+        raise ValueError(f"Expected 1D mask array - got array with shape {mask.shape}")
+
+    # Construct second-order difference matrix
+    D = np.array([[1, -2, 1]]).T.repeat(N - 1, axis=1)
+    D = dia_array((D, [-2, -1, 0]), shape=(N, N - 2))
+    Hp = scale * D @ D.T
+
+    # Create the banded smoothness matrix and weights matrix
+    H = np.ones((3, N), dtype=np.float64)
+    W = np.zeros_like(H)
+
+    # Fill the lower banded matrix
+    for i in range(H.shape[0]):
+        H[i, : N - i] = Hp.diagonal(i)
+
+    # Initialize weights to one
+    W[0] = 1.0
+
+    # Get the maximum exponential to avoid runtime warnings
+    maxpwr = np.log(np.finfo(y.dtype).max)
+
+    for _ in range(max_iter):
+        # Ignore masked values
+        W[:, mask] = 0.0
+        # Extract the actual weights. W is a 3xN matrix to match
+        # the banded shape of H, all off-diagonal elements are zero
+        w = W[0]
+
+        z = la.solveh_banded(H + W, w * y, lower=True)
+
+        # Get the difference between the signal and the baseline estimate,
+        # and compute the meand a std where unmasked data is less than zero
+        d = y - z
+        dn = d[(d < 0) & ~mask]
+        m = np.mean(dn)
+        s = np.std(dn)
+
+        # Adjust weights based on the criteria discussed in the paper
+        pwr = 2 * (d - ((2 * s) - m)) * invert_no_zero(s)
+        pwr = np.clip(pwr, -maxpwr, maxpwr)
+        wt = invert_no_zero(1 + np.exp(pwr))
+
+        # Check for convergeance
+        if la.norm(w - wt) / la.norm(w) < end_frac:
+            break
+
+        # Update the weights
+        W[0] = wt
+
+    return z
+
+
+def taper_mask(mask, nwidth, outer=False):
+    """Taper a 2d mask along the last axis.
+
+    Parameters
+    ----------
+    mask : np.ndarray
+        Mask to taper
+    nwidth : int
+        Width of the taper window
+    outer : bool, optional
+        If True, expand the mask outwards (wider). Otherwise,
+        expand the mask inwards (narrower). Default is False
+
+    Returns
+    -------
+    tapered_mask : np.ndarray[np.float64]
+        Mask convolved with taper window
+    """
+    mask = np.atleast_2d(mask)
+
+    width = 2 * nwidth - 1
+
+    taper = np.hanning(width)[np.newaxis]
+    taper /= np.sum(taper)
+
+    tapered_mask = np.zeros(
+        (mask.shape[0], mask.shape[-1] + 2 * width), dtype=np.float64
+    )
+    tapered_mask[:, width:-width] = mask.astype(np.float64)
+    # Extend the edges
+    tapered_mask[:, :width] = tapered_mask[:, width][:, np.newaxis]
+    tapered_mask[:, -width:] = tapered_mask[:, -width - 1][:, np.newaxis]
+
+    if outer:
+        tapered_mask = 1.0 - tapered_mask
+
+    # First convolution
+    tapered_mask = np.isclose(
```

Comment:
The purpose of the `isclose` line isn't obvious to me on first reading. A comment might be helpful.

Reply:
I'm still confused about this. If the input mask is just ones and zeros, applying the taper will result in array with numbers <1, which `np.allclose` will uniformly set to zero. What am I missing here?

Reply:
Ok I'm looking at it again and I've confused myself. I think I need to poke around in a notebook and figure out what I was doing

Reply:
I've gone back through to better understand why this is done and clarified it for myself.

The idea is that the taper is always applied to the inside of the `True` regions during the convolution - i.e., any value that is 0 before convolving with the kernel will remain 0 after, and then the `1 - x` is applied afterwards if the taper should be outside. The first convolution effectively narrows the region where `x == 1` by `nwidth`, but leaves a taper which extends by `nwidth` outside the desired boundary. So, by doing `np.isclose(..., 1)` and then applying a second convolution, the taper is applied at the correct width. There are a couple of plots below to show what's happened
![taper_step_1](https://github.com/radiocosmology/draco/assets/22264047/a7b91020-97de-4c25-bfbf-7f70f97059ff)
![taper_step_2](https://github.com/radiocosmology/draco/assets/22264047/fb0f0a56-e311-40d7-a442-9785e90e2063)

Reply:
Got it, thanks for the extra explanation!